ext/js/language/translator.js

```diff
@@ -497,36 +469,59 @@ export class Translator {
     }
 
     /**
-     * @param {import('language').TextProcessorWithId<unknown>[]} textProcessors
-     * @param {import('translation-internal').TextProcessorVariant} processorVariant
      * @param {string} text
+     * @param {import('language').TextProcessorWithId<unknown>[]} textProcessors
+     * @param {(import('translation').FindTermsTextReplacement[] | null)[]} textReplacements
      * @param {import('translation-internal').TextCache} textCache
-     * @returns {string}
-     */
-    _applyTextProcessors(textProcessors, processorVariant, text, textCache) {
-        for (const {id, textProcessor: {process}} of textProcessors) {
-            const setting = processorVariant.get(id);
-
-            let level1 = textCache.get(text);
-            if (!level1) {
-                level1 = new Map();
-                textCache.set(text, level1);
+     * @returns {Set<string>}
+     */
+    _getTextVariants(text, textProcessors, textReplacements, textCache) {
+        let variants = new Set([text]);
+        for (const textReplacement of textReplacements) {
+            if (textReplacement === null) { continue; }
+            variants.add(this._applyTextReplacements(text, textReplacement));
+        }
+        for (const {id, textProcessor: {process, options}} of textProcessors) {
+            /** @type {Set<string>} */
+            const newVariants = new Set();
+            for (const variant of variants) {
+                for (const option of options) {
+                    const processed = this._getProcessedText(textCache, variant, id, option, process);
+                    newVariants.add(processed);
+                }
             }
+            variants = newVariants;
```

Comment:
doesn't `variants` get overshadowed every time this for-loop runs?

Reply:
It's hard to explain these things, but essentially we are starting with a set that contains just the raw word, and then for each cycle of the loop we go one step deeper on the tree, replacing that word with every variant given by the current preprocessor.

To borrow the example from the docs for english we go

1. `['reaD']`
2. `['reaD', 'read']` (variants for `decapitalize`)
3. `['reaD', 'ReaD', 'read', 'Read'] (variants for each of the last two for `capitalizeFirstLetter`)

You can see how they form a tree, but we only need to remember the current layer.